File: lecture02/02_anagrams/anagrams.cpp
// anagrams.cpp

#include <iostream>
#include <vector>

using namespace std;

// Constants

const size_t NLETTERS = 26;

// Functions

bool is_anagram(string &s, string &t) {
    size_t sc[NLETTERS] = {0};
    size_t tc[NLETTERS] = {0};

    for (auto c : s) { sc[tolower(c) - 'a']++; }
    for (auto c : t) { tc[tolower(c) - 'a']++; }

    for (size_t i = 0; i < NLETTERS; i++) {
    	if (sc[i] != tc[i]) {
    	    return false;
	}
    }

    return true;
}

// Main execution

int main(int argc, char *argv[]) {
    string first, second;

    while (cin >> first >> second) {
    	cout << (is_anagram(first, second) ? "Anagram" : "Not Anagram") << endl;
    }

    return 0;
}

File: lecture02/02_anagrams/anagrams.py
#!/usr/bin/env python3

import collections
import sys

# Functions

def is_anagram_sorted(s, t):
    return sorted(s.lower()) == sorted(t.lower())

def is_anagram_histogram(s, t):
    sc = collections.Counter(s.lower())
    tc = collections.Counter(t.lower())
    return sc == tc

is_anagram = is_anagram_histogram

# Main Execution

if __name__ == '__main__':
    for line in sys.stdin:
        first, second = line.strip().split()
        print('Anagram' if is_anagram(first, second) else 'Not Anagram')
